File: test_filtro_imagens.js
/**
 * Testa o filtro JS do fechar_os.py contra HTML que replica
 * a estrutura REAL do AWO-Soft (suporteprime.awo-soft.com)
 */
const { JSDOM } = require('jsdom');
const fs = require('fs');

const src = fs.readFileSync('fechar_os.py', 'utf8');
const match = src.match(/img_urls = page\.evaluate\("""\s*([\s\S]*?)\s*"""\s*\) or \[\]/) ||
              src.match(/img_urls = page\.evaluate\("""\s*([\s\S]*?)\s*"""\) or \[\]/);
if (!match) { console.error('Não encontrei o bloco JS'); process.exit(1); }
const JS_FILTRO = match[1];

// HTML que replica exactamente a estrutura do AWO-Soft visível no screenshot
const html = `<!DOCTYPE html><html><body>

  <!-- NAVBAR topo (logo AWO deve ser EXCLUÍDA) -->
  <nav class="navbar topbar">
    <img id="logo-awo"
         src="https://suporteprime.awo-soft.com/assets/img/logo-awo.png"
         data-w="120" data-h="40">
  </nav>

  <!-- SIDEBAR esquerda (ícones de navegação, devem ser EXCLUÍDOS) -->
  <aside class="sidebar">
    <img id="icon-dashboard"
         src="https://suporteprime.awo-soft.com/assets/icons/dashboard.svg"
         data-w="24" data-h="24">
    <img id="icon-clientes"
         src="https://suporteprime.awo-soft.com/assets/icons/clients.svg"
         data-w="24" data-h="24">
  </aside>

  <!-- CONTEÚDO PRINCIPAL -->
  <main>
    <!-- Tabs: Ficha | Periodicidade | Serviços | Custos | Imagens/Doc. | Intervenções -->
    <div id="tab_images">

      <!-- Secção "Imagens e Documentos" (estrutura real do AWO) -->
      <h2>Imagens e Documentos</h2>

      <div class="images-grid">
        <!-- 3 fotos reais de trabalho (devem ser INCLUÍDAS) -->
        <div>
          <p>Nome: .</p>
          <img id="foto1"
               src="https://suporteprime.awo-soft.com/storage/work-orders/14437/foto1.jpg"
               data-w="800" data-h="600">
        </div>
        <div>
          <p>Nome: .</p>
          <img id="foto2"
               src="https://suporteprime.awo-soft.com/storage/work-orders/14437/foto2.jpg"
               data-w="1200" data-h="900">
        </div>
        <div>
          <p>Nome: .</p>
          <img id="foto3"
               src="https://suporteprime.awo-soft.com/storage/work-orders/14437/foto3.jpg"
               data-w="1024" data-h="768">
        </div>
      </div>

    </div>
  </main>

</body></html>`;

const dom = new JSDOM(html, { runScripts: 'dangerously', url: 'https://suporteprime.awo-soft.com/work-orders/view/14437#tab_images' });
const { document: doc, Object: Obj } = dom.window;

// Simula naturalWidth/naturalHeight
doc.querySelectorAll('img').forEach(img => {
  const w = parseInt(img.getAttribute('data-w') || '0');
  const h = parseInt(img.getAttribute('data-h') || '0');
  Obj.defineProperty(img, 'naturalWidth',  { get: () => w, configurable: true });
  Obj.defineProperty(img, 'naturalHeight', { get: () => h, configurable: true });
});

const resultado = dom.window.eval(`(${JS_FILTRO})()`);

console.log('\n=== Teste filtro imagens AWO (estrutura real) ===\n');
console.log(`  Imagens na página : 5 (1 logo navbar + 2 ícones sidebar + 3 fotos)`);
console.log(`  Retornadas         : ${resultado.length}`);
resultado.forEach(u => console.log(`    ✓ ${u.split('/').slice(-2).join('/')}`));

let ok = true;

const temLogo = resultado.some(u => u.includes('logo-awo') || u.includes('logo'));
if (temLogo) { console.log('\n  ✗ FALHA: logo AWO incluída!'); ok = false; }
else          { console.log('\n  ✓ Logo AWO excluída (navbar)'); }

const temIcones = resultado.some(u => u.includes('icons/'));
if (temIcones) { console.log('  ✗ FALHA: ícones sidebar incluídos!'); ok = false; }
else           { console.log('  ✓ Ícones sidebar excluídos'); }

const fotos = resultado.filter(u => u.includes('storage/work-orders'));
if (fotos.length !== 3) {
  console.log(`  ✗ FALHA: esperadas 3 fotos, obtidas ${fotos.length}`);
  ok = false;
} else {
  console.log(`  ✓ ${fotos.length} fotos de trabalho incluídas`);
}

console.log(ok ? '\n  TODOS OS TESTES PASSARAM ✓\n' : '\n  ALGUNS TESTES FALHARAM ✗\n');
process.exit(ok ? 0 : 1);

File: test_filtro_imagens.py
"""
Testa o filtro JS de imagens do fechar_os.py contra uma página HTML
que replica a estrutura real do AWO-Soft:
  - logo AWO (quadrada, azul) na sidebar  →  deve ser EXCLUÍDA
  - fotos de trabalho no painel de imagens →  devem ser INCLUÍDAS
"""

import re
from playwright.sync_api import sync_playwright

# ── Extrai o bloco JS do fechar_os.py ──────────────────────────────────
def extrair_js_filtro() -> str:
    src = open("fechar_os.py", encoding="utf-8").read()
    start = src.find("img_urls = page.evaluate(")
    end   = src.find('    """) or []', start) + len('    """) or []')
    bloco = src[start:end]
    # Extrai só a função JS entre as aspas triplas
    m = re.search(r'page\.evaluate\("""\s*(.*?)\s*"""\)', bloco, re.DOTALL)
    if not m:
        raise RuntimeError("Não encontrei o bloco JS no fechar_os.py")
    return m.group(1)

JS_FILTRO = extrair_js_filtro()

# ── HTML que replica estrutura AWO ──────────────────────────────────────
HTML_AWO = """<!DOCTYPE html>
<html>
<head><title>AWO Mock</title></head>
<body>

<!-- SIDEBAR com logo AWO-Soft (deve ser EXCLUÍDA) -->
<div class="sidebar">
  <div class="brand logo">
    <!-- Logo AWO: quadrada 200x200, sem keyword no URL -->
    <img id="logo-awo"
         src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMgAAADICAYAAACtWK6eAAAABmJLR0QA/wD/AP+gvaeTAAAADklEQVQI12NgGAWkAgABNgABzKyFIQAAAABJRU5ErkJggg=="
         width="200" height="200"
         style="width:200px;height:200px">
  </div>
</div>

<!-- CONTEÚDO PRINCIPAL -->
<main>
  <!-- Tab de imagens activa -->
  <div class="tab-pane active" id="tab-imagens">

    <!-- Fotos reais do trabalho (devem ser INCLUÍDAS) -->
    <div class="gallery">
      <img id="foto1"
           src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEASABIAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/wAARC AACAAIDAQACIABFARABEQACIABFARAEQACIAB"
           width="800" height="600" style="width:800px;height:600px">
      <img id="foto2"
           src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEASABIAAD/2wBDABBoto"
           width="1200" height="900" style="width:1200px;height:900px">
      <img id="foto3"
           src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEASABIAAD/2wBDABFoto3"
           width="1024" height="768" style="width:1024px;height:768px">
    </div>

  </div>
</main>

<script>
  // Simula dimensões naturais (no browser real vêm do ficheiro)
  document.querySelectorAll('img').forEach(img => {
    const w = parseInt(img.style.width);
    const h = parseInt(img.style.height);
    Object.defineProperty(img, 'naturalWidth',  {get: () => w});
    Object.defineProperty(img, 'naturalHeight', {get: () => h});
  });
</script>
</body>
</html>"""


def run_tests():
    print("\n=== Teste do filtro de imagens AWO ===\n")
    falhas = []

    with sync_playwright() as pw:
        browser = pw.chromium.launch(headless=True)
        page    = browser.new_page()
        page.set_content(HTML_AWO)
        page.wait_for_timeout(500)

        # Corre o JS do fechar_os.py
        resultado = page.evaluate(JS_FILTRO)
        browser.close()

    print(f"  Imagens retornadas pelo filtro: {len(resultado)}")
    for url in resultado:
        print(f"    • {url[:80]}...")

    # ── Verificações ────────────────────────────────────────────────────
    ids_retornados = resultado  # são data URIs, verificamos por exclusão

    # 1. Logo AWO (200x200 quadrada no sidebar) deve estar excluída
    # O logo tem 200x200 → área 40000, ratio 1.0 → deve ser excluído pela regra de logo ≤300
    # Verificamos indiretamente: o número de imagens deve ser < total da página
    total_na_pagina = 4  # logo + 3 fotos
    if len(resultado) >= total_na_pagina:
        falhas.append(f"FALHA: filtro retornou {len(resultado)} imagens (esperado < {total_na_pagina})")
    else:
        print(f"\n  ✓ Filtro excluiu {total_na_pagina - len(resultado)} imagem(ns) (logo/UI)")

    # 2. Fotos reais devem ser incluídas (800x600, 1200x900, 1024x768)
    if len(resultado) < 1:
        falhas.append("FALHA: nenhuma foto de trabalho foi incluída")
    else:
        print(f"  ✓ {len(resultado)} foto(s) de trabalho incluída(s)")

    # 3. Zero imagens com dimensão ≤ 300x300 quadradas devem passar
    # (simulamos verificando que o logo de 200x200 foi excluído)
    # O logo é o 1º item, as fotos são maiores que 300px
    fotos_grandes = [u for u in resultado if u]  # todas as retornadas são fotos
    print(f"\n  Resultado: {len(resultado)} foto(s) incluídas, {total_na_pagina - len(resultado)} excluídas")

    if falhas:
        print("\n  FALHAS:")
        for f in falhas:
            print(f"    ✗ {f}")
        return False
    else:
        print("\n  TODOS OS TESTES DO FILTRO PASSARAM ✓")
        return True


if __name__ == "__main__":
    ok = run_tests()
    exit(0 if ok else 1)
